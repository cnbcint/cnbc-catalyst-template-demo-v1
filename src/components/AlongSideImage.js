import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";
import Image from "./Image";

import { breakpoints, getMargin } from "../helpers/Layout";

const ImageContainer = styled.div`
  @media (min-width: ${breakpoints.mobile}) {
    float: ${(props) => props.position};
    display: block;
    margin-right: ${(props) =>
      props.position === "left" ? getMargin(2) : "auto"};
    margin-left: ${(props) =>
      props.position === "right" ? getMargin(2) : "auto"};
    width: 50%;
    img {
      margin: 0;
      margin-top: ${getMargin(1)};
      margin-bottom: ${getMargin(1)};
    }
  }
`;

/**
 * Add an image capable of be wrapper around text left or right of the text
 */
export default function AlongSideImage(props) {
  return (
    <div>
      <ImageContainer position={props.position}>
        <Image source={props.image} alt={props.alt} />
      </ImageContainer>
      {props.children}
    </div>
  );
}

AlongSideImage.defaultProps = {
  position: "left",
};

AlongSideImage.propTypes = {
  /**
   * Allows 2 types of alignment
   */
  position: PropTypes.oneOf(["left", "right"]),

  /**
   * Image object with values for mobile, tablet and desktop
   */
  image: PropTypes.shape({
    mobile: PropTypes.string,
    tablet: PropTypes.string,
    desktop: PropTypes.string,
  }).isRequired,

  /**
   * Image alternate text
   */
  alt: PropTypes.string.isRequired,

  /**
   * The content
   */
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,
};
