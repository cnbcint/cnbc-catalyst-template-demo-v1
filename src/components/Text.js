import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const Pdrop = styled.p`
  color: ${(props) => props.color};
  font-size: ${(props) => (props.highlight ? "1.4em" : "1.2rem")};
  line-height: ${(props) => (props.highlight ? "140%" : "125%")};
  text-align: ${(props) => props.textAlign};
  max-width: 600px;
  margin:0 auto;
  &::first-letter {
    color: ${(props) => props.dropCapColor};
    padding: 0 0;
    margin: 0.5rem 1rem 0 0;
    font-size: 5.5rem;
    font-weight: bold;
    float: left;
    line-height: 4rem;
  }
`;
const P = styled.p`
  color: ${(props) => props.color};
  font-size: ${(props) => (props.highlight ? "1.4rem" : "1.2rem")};
  line-height: ${(props) => (props.highlight ? "140%" : "125%")};
  text-align: ${(props) => props.textAlign};
  max-width: 600px;
  margin:0 auto;

`;
/** 
 Text Block will add a paragraph element to it's children. Do not add P inside this element to avoid double nested P element
*/
export default function Text(props) {
  if (props.dropCap) {
    return (
      <Pdrop
        highlight={props.highlight}
        dropCapColor={props.dropCapColor ? props.dropCapColor : props.color}
        color={props.color}
        textAlign={props.textAlign}
        style={props.style}
        className={props.className}
      >
        {props.children}
      </Pdrop>
    );
  }
  return (
    <P
      highlight={props.highlight}
      color={props.color}
      style={props.style}
      textAlign={props.textAlign}
      className={props.className}
    >
      {props.children}
    </P>
  );
}

Text.defaulProps = {
  dropCap: false,
  color: "inherit",
  dropCapColor: "",
  highlight: false,
  className: "",
  textAlign: "left",
};

Text.propTypes = {
  /**
   Will Increase font size to highlight the text
   */
  highlight: PropTypes.bool,
  /**
   Will Increase font size of the first letter of the paragraph to achive DropCap 
   */
  dropCap: PropTypes.bool,
  /**
   CSS override for font color
   */
  color: PropTypes.string,
  /**
   CSS text aligment
   */
  textAlign: PropTypes.oneOf(["left", "right", "center"]),
  /**
   CSS color for drop cap
   */
  dropCapColor: PropTypes.string,
  /**
  React CSS format to override rules has required by design
  */
  style: PropTypes.object,
  className: PropTypes.string,
  children: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.array,
    PropTypes.element,
  ]).isRequired,
};
