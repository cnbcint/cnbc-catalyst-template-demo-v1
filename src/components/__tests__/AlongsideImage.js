import React from "react";
import AlongSideImage from "../AlongSideImage";
import Text from "../Text";
import { render } from "@testing-library/react";
import "jest-styled-components";

const Component = () => (
  <AlongSideImage
    position="right"
    image={{
      desktop:
        "http://localhost:8080/images/responsive/output/image02-desktop.jpg",
      tablet:
        "http://localhost:8080/images/responsive/output/image02-tablet.jpg",
      mobile:
        "http://localhost:8080/images/responsive/output/image02-mobile.jpg",
    }}
    alt="Test Image"
  >
    <Text>
      The standard chunk of Lorem Ipsum used since the 1500s is reproduced below
      for those interested. Sections 1.10.32 and 1.10.33 from &quot;de Finibus
      Bonorum et Malorum&quot; by Cicero are also reproduced in their exact
      original form, accompanied by English versions from the 1914 translation
      by H. Rackham.
    </Text>
  </AlongSideImage>
);

beforeAll(() => {
  window.IntersectionObserver = class {
    observe() {
      return null;
    }
    disconnect() {}
  };
});

describe("AlongSideImage", () => {
  it("Renders correctly", () => {
    const { container } = render(<Component />);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("Image is rendered", () => {
    const { container } = render(<Component />);
    expect(container.getElementsByTagName("picture").length === 1).toBe(true);
  });

  it("Text is rendered", () => {
    const { container } = render(<Component />);
    expect(container.getElementsByTagName("p").length === 1).toBe(true);
  });
});

afterAll(() => {
  window.IntersectionObserver = null;
});
