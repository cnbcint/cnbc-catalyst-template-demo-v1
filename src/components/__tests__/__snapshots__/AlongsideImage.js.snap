// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AlongSideImage Renders correctly 1`] = `
.c1 {
  width: 100%;
}

.c2 {
  font-size: 1.2rem;
  line-height: 125%;
}

@media (min-width:768px) {
  .c0 {
    float: right;
    display: block;
    margin-right: auto;
    margin-left: 2rem;
    width: 50%;
  }

  .c0 img {
    margin: 0;
    margin-top: 1rem;
    margin-bottom: 1rem;
  }
}

<div>
  <div
    class="c0"
  >
    <picture
      class=""
    >
      <source
        media="(min-width:1280px)"
        srcset="http://localhost:8080/images/responsive/output/image02-desktop.jpg"
      />
      <source
        media="(min-width:768px)"
        srcset="http://localhost:8080/images/responsive/output/image02-tablet.jpg"
      />
      <img
        alt="Test Image"
        class="c1 pre-animated"
        loading="lazy"
        src="http://localhost:8080/images/responsive/output/image02-mobile.jpg"
      />
    </picture>
  </div>
  <p
    class="c2"
  >
    The standard chunk of Lorem Ipsum used since the 1500s is reproduced below for those interested. Sections 1.10.32 and 1.10.33 from "de Finibus Bonorum et Malorum" by Cicero are also reproduced in their exact original form, accompanied by English versions from the 1914 translation by H. Rackham.
  </p>
</div>
`;
